### DISCUSSION TITLE
paper.fitToContent and baseWidth do not work as expected

### DISCUSSION THREAD

**Discussion by mwanagosos** [clientIO/joint#1680] [Q&A]
I'm using JointJS+ 3.5.
I have a paper that's very big.
```
new dia.Paper({
  width: 6000,
  height: 6000,
  // ...
});
```

I want to make it a lot smaller

```
new dia.Paper({
  width: 150,
  height: 150,
  // ...
});
```

and implement 
```
this.app.paperScroller = new ui.PaperScroller({
  autoResizePaper: true,
  scrollWhileDragging: true,
  baseHeight: 1500,
  baseWidth: 1500,
  // ...
});
```
on the paper scroller with `autoResizePaper` and `scrollWhileDragging`.

I want the initial size of the graph to be 1500x1500. When an element is dragged outside of the paper, I want it to stretch by 150 only.
Unfortunately, it dragging an element outside of a paper makes it bigger by 1500 instead.
It happens even though the documentation states to use `baseWidth` and `baseHeight` if we want to auto-resize the paper by a value different that the `width` and `height` set in the Paper constructor.

Also, I have a lot of old JSON files with exported graphs. I still want to support them.
I'm doing
```
paper.fitToContent();
```
to crop the paper on the graphs imported through `graph.fromJSON`.
It seems to work fine until I move one of the blocks or call `paperScroller.zoomToFit()`. As soon as I do that, paper becomes very big again and `paper.fitToContent()` gets reverted.

Is there something I'm missing?

**Comment by kumilingus** [comment]
The `PaperScroller` view calls `paper.fitToContent()` internally. You can control the resizing with `contentOptions` option.

```js
new ui.PaperScroller({

    // This is the size increment.
    // If one paper is not enough to cover the whole graph,
    // you lay more papers next to it until the whole graph fits. This one paper is the base.
    // Note: it's an equivalent of `gridWidth` and `gridHeight` of `fitToContent()` method
    baseWidth: 150,
    baseHeight: 150,

    contentOptions: {
        // This is the minimum paper size regardless of content size.
        minWidth: 1500,
        minHeight: 1500,
        
        /* ... any other option from fitToContent() */
    }
    
    /* ... any other option of PaperScroller */
});
```

**Reply by mwanagosos** [reply]
@kumilingus I'm still experiencing the bug where `paper.fitToContent()` gets reverted as soon as I move one of the blocks, or use `paperScroller.zoomToFit()`. What would you suggest to try?

**Reply by kumilingus** [reply]
If `autoResizePaper` is enabled, any change made to the graph results in calling `paper.fitToContent()` with `contentOptions`. You would have to make sure the `paper.fitToContent()` options and PaperScroller's `contentOptions` are in sync. However, it is better not to use `paper.fitToContent()` at all and let the PaperScroller resize it with `paperScroller.adjustPaper()`.

If above won't help you have to provide us with an example.

---

If your paper is in `async` mode, make sure to use `useModelGeometry` option (everywhere).

**Reply by mwanagosos** [reply]
I'm doing the following:
```
new ui.PaperScroller({
  paper: this.paper,
  cursor: 'grab',
  autoResizePaper: true,
  scrollWhileDragging: true,
  baseHeight: 500,
  baseWidth: 500,
  contentOptions: {
    minWidth: 1500,
    minHeight: 1500,
    allowNewOrigin: 'negative',
    useModelGeometry: true,
  }
});
```
and
```
new dia.Paper({
  model: this.graph,
  width: 500,
  height: 500,
  gridSize: 30,
  background: {
    color: '#ffffff',
  },
  drawGrid: {
    name: 'dot',
    thickness: 2,
    color: '#969696',
  },
  cellViewNamespace: cellNamespace,
  snapLinks: { radius: 75 },
  validateConnection(
    sourceCellView,
    sourceMagnet,
    targetCellView,
    targetMagnet,
  ) {
   // ...
  },
  validateMagnet(cellView, magnet) {
   // ...
  },
  defaultLink: new dia.Link({
    // ...
  }),
});
```
Using `paperScroller.adjustPaper()` does not seem to do anything, unfortunately.
I've imported a graph that was created on a 6000x6000 paper and contains just a few blocks. Then, I called
```
paperScroller.centerContent();
paperScroller.adjustPaper();
paperScroller.zoomToFit({ padding: 20 });
```
but the content isn't centered and the paper isn't adjusted.

`paper.options.width` is `9107.3228297144`
`paper.options.height` is `7968.907476000099`
which is way to big for the imported graph.

**Reply by mwanagosos** [reply]
![Screenshot from 2022-05-27 20-15-26](https://user-images.githubusercontent.com/80329180/170767811-24349d37-b4a4-4dd9-aa6c-44cacac4bb89.png)

**Reply by kumilingus** [reply]
I think `minWidth` and `minHeight` do not play well with paper origin other than `0,0`. You want to set the bottom-right corner of the paper to minimum of { x: 1500, y: 1500 } (and not just the size of the paper).

You can resize the paper based on a union of the graph rectangle and minimal graph area.

```js
const paperScroller = new joint.ui.PaperScroller({
    paper,
    autoResizePaper: true,
    baseHeight: 500,
    baseWidth: 500,
    contentOptions: (paperScroller) => {
        const minSize = 1500;
        return {
            contentArea: g.Rect.fromRectUnion(graph.getBBox(), new g.Rect(0, 0, minSize, minSize)),
        }
    }
});
```

This works perfectly for me in the KitchenSink application. If it does not work for you, please provide steps how to reproduce the issue in any of our applications or send us an example.

**Reply by mwanagosos** [reply]
@kumilingus I think I didn't understand the `paperScroller.centerContent` function correctly. I thought that it would reposition the blocks to be in the center of the paper. It looks like it just scrolls the paper so that the content appears in the center of the view.
Is there a function that would reposition all of the blocks so that the move to the center of the actual paper?

**Reply by kumilingus** [reply]
The paperScroller never modifies the graph (positions of elements). It only changes the transformations (scale, translate) of the graph view i.e paper.

To move all cells to the middle of the current paper area use the code below.
```js
const { x, y } =  paper.getArea().center().difference(graph.getBBox().center());
graph.translate(x, y);
```

**Reply by mwanagosos** [reply]
@kumilingus Thank you. This was the missing piece that helped me achieve the behavior I needed :+1:

**Reply by mwanagosos** [reply]
By the way, I've been working on auto-resizing the paper to benefit more from having the Navigator.
Unfortunately, I've noticed a significant performance hit when using the Navigator with some of the bigger graphs. Is there a way to counter that?

**Reply by kumilingus** [reply]
@mwanagosos If the question is not related to #1680 please open a new discussion.

**Reply by mwanagosos** [reply]
I did that. Thanks a lot.
